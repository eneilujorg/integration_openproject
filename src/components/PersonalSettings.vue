<template>
	<div id="openproject_prefs" class="section">
		<SettingsTitle />
		<div id="openproject-content">
			<div id="toggle-openproject-navigation-link">
				<input
					id="openproject-link"
					type="checkbox"
					class="checkbox"
					:checked="state.navigation_enabled"
					@input="onNavigationChange">
				<label for="openproject-link">{{ t('integration_openproject', 'Enable navigation link') }}</label>
			</div>
			<br><br>
			<div v-if="connected" class="openproject-grid-form">
				<label class="openproject-connected">
					<a class="icon icon-checkmark-color" />
					{{ t('integration_openproject', 'Connected as {user}', { user: state.user_name }) }}
				</label><br>
				<button id="openproject-rm-cred" @click="onLogoutClick">
					<span class="icon icon-close" />
					{{ t('integration_openproject', 'Disconnect from OpenProject') }}
				</button>
			</div>
			<div v-if="connected" id="openproject-search-block">
				<input
					id="search-openproject"
					type="checkbox"
					class="checkbox"
					:checked="state.search_enabled"
					@input="onSearchChange">
				<label for="search-openproject">{{ t('integration_openproject', 'Enable unified search for tickets') }}</label>
				<br><br>
				<p v-if="state.search_enabled" class="settings-hint">
					<span class="icon icon-details" />
					{{ t('integration_openproject', 'Warning, everything you type in the search bar will be sent to your OpenProject instance.') }}
				</p>
				<input
					id="notification-openproject"
					type="checkbox"
					class="checkbox"
					:checked="state.notification_enabled"
					@input="onNotificationChange">
				<label for="notification-openproject">{{ t('integration_openproject', 'Enable notifications for activity in my work packages') }}</label>
			</div>
			<OAuthConnectButton v-else :request-url="requestUrl" :admin-config-status="adminConfigStatus" />
		</div>
	</div>
</template>

<script>
import { loadState } from '@nextcloud/initial-state'
import { generateUrl } from '@nextcloud/router'
import axios from '@nextcloud/axios'
import { showSuccess, showError } from '@nextcloud/dialogs'
import '@nextcloud/dialogs/styles/toast.scss'
import SettingsTitle from '../components/settings/SettingsTitle'
import OAuthConnectButton from './OAuthConnectButton'

export default {
	name: 'PersonalSettings',

	components: {
		SettingsTitle, OAuthConnectButton,
	},

	props: [],

	data() {
		return {
			loading: false,
			state: loadState('integration_openproject', 'user-config'),
			requestUrl: loadState('integration_openproject', 'user-config').request_url,
			adminConfigStatus: loadState('integration_openproject', 'admin-config-status'),
		}
	},

	computed: {
		connected() {
			if (!this.adminConfigStatus) return false
			return this.state.token && this.state.token !== ''
				&& this.state.user_name && this.state.user_name !== ''
		},
	},

	mounted() {
		const paramString = window.location.search.substr(1)
		// eslint-disable-next-line
		const urlParams = new URLSearchParams(paramString)
		const zmToken = urlParams.get('openprojectToken')
		if (zmToken === 'success') {
			showSuccess(t('integration_openproject', 'Successfully connected to OpenProject!'))
		} else if (zmToken === 'error') {
			showError(t('integration_openproject', 'OAuth access token could not be obtained:') + ' ' + urlParams.get('message'))
		}
	},

	methods: {
		onLogoutClick() {
			this.state.token = ''
			this.saveOptions({ token: this.state.token, token_type: '' })
		},
		onNotificationChange(e) {
			this.state.notification_enabled = e.target.checked
			this.saveOptions({ notification_enabled: this.state.notification_enabled ? '1' : '0' })
		},
		onSearchChange(e) {
			this.state.search_enabled = e.target.checked
			this.saveOptions({ search_enabled: this.state.search_enabled ? '1' : '0' })
		},
		onNavigationChange(e) {
			this.state.navigation_enabled = e.target.checked
			this.saveOptions({ navigation_enabled: this.state.navigation_enabled ? '1' : '0' })
		},
		saveOptions(values) {
			const req = {
				values,
			}
			const url = generateUrl('/apps/integration_openproject/config')
			axios.put(url, req)
				.then((response) => {
					showSuccess(t('integration_openproject', 'OpenProject options saved'))
					if (response.data.user_name !== undefined) {
						this.state.user_name = response.data.user_name
						if (this.state.token && response.data.user_name === '') {
							showError(t('integration_openproject', 'Incorrect access token'))
						}
					}
				})
				.catch((error) => {
					console.debug(error.response)
					const msg = error.response?.data?.errorMessage === 'Invalid token'
						? t('integration_openproject', 'Invalid token')
						: error.response?.data?.errorMessage === 'Not found'
							? t('integration_openproject', 'OpenProject instance not found')
							: error.response?.request?.responseText
					showError(
						t('integration_openproject', 'Failed to save OpenProject options')
						+ ': ' + msg
					)
				})
				.then(() => {
					this.loading = false
				})
		},
	},
}
</script>

<style scoped lang="scss">
#openproject-search-block {
	margin-top: 30px;
}

.openproject-grid-form label {
	line-height: 38px;
}

.openproject-grid-form input {
	width: 100%;
}

.openproject-grid-form {
	max-width: 600px;
	display: grid;
	grid-template: 1fr / 1fr 1fr;
	button .icon {
		margin-bottom: -1px;
	}
}

#openproject_prefs .icon {
	display: inline-block;
	width: 32px;
}

#openproject_prefs .grid-form .icon {
	margin-bottom: -3px;
}

#openproject-content {
	margin-left: 40px;
}

#openproject-search-block .icon {
	width: 22px;
}

#openproject-content .oauth-connect--message {
  text-align: left !important;
}
</style>
