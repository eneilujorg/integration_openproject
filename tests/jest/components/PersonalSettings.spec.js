/* jshint esversion: 8 */

import { shallowMount, createLocalVue } from '@vue/test-utils'
import PersonalSettings from '../../../src/components/PersonalSettings.vue'
import * as initialState from '@nextcloud/initial-state'

const localVue = createLocalVue()

// eslint-disable-next-line no-import-assign
initialState.loadState = jest.fn((key, value) => {
	if (value === 'user-config') {
		return {
			request_url: 'https://nextcloud.com/',
		}
	} else if (value === 'admin-config-status') return false
})

describe('PersonalSettings.vue Test', () => {
	describe('oAuth', () => {
		const oAuthButtonSelector = 'oauthconnectbutton-stub'
		const oAuthDisconnectButtonSelector = '#openproject-rm-cred'
		const connectedAsLabenSelector = '.openproject-connected'
		const searchBlockSelector = '#openproject-search-block'
		let wrapper
		beforeEach(() => {
			wrapper = shallowMount(PersonalSettings, {
				localVue,
				mocks: {
					t: (app, msg) => msg,
					generateUrl() {
						return '/'
					},
				},
			})
		})

		describe('when admin config status is ok', () => {
			describe.each([
				{ user_name: 'test', token: '' },
				{ user_name: 'test', token: null },
				{ user_name: 'test', token: undefined },
				{ user_name: '', token: '123' },
				{ user_name: null, token: '123' },
				{ user_name: undefined, token: '123' },
				{ user_name: '', token: '' },
				{ user_name: null, token: '' },
				{ user_name: '', token: null },
			])('when username or token not given', (cases) => {
				beforeEach(async () => {
					await wrapper.setData({
						adminConfigStatus: true,
						state: cases,
					})
				})
				it('oAuth connect button is displayed', () => {
					expect(wrapper.find(oAuthButtonSelector).exists()).toBeTruthy()
				})
				it('search settings are not displayed', () => {
					expect(wrapper.find(searchBlockSelector).exists()).toBeFalsy()
				})
				it('oAuth disconnect button is not displayed', () => {
					expect(wrapper.find(oAuthDisconnectButtonSelector).exists()).toBeFalsy()
				})
			})
			describe('when username and token are given', () => {
				beforeEach(async () => {
					await wrapper.setData({
						adminConfigStatus: true,
						state: { user_name: 'test', token: '123' },
					})
				})
				it('oAuth connect button is not displayed', () => {
					expect(wrapper.find(oAuthButtonSelector).exists()).toBeFalsy()
				})
				it('oAuth disconnect button is displayed', () => {
					expect(wrapper.find(oAuthDisconnectButtonSelector).exists()).toBeTruthy()
				})
				it('connected as label is displayed', () => {
					expect(wrapper.find(connectedAsLabenSelector).text()).toBe('Connected as {user}')
				})
				it('oAuth search settings are displayed', () => {
					expect(wrapper.find(searchBlockSelector).exists()).toBeTruthy()
				})
			})
		})
		describe('when admin config status is not ok', () => {
			beforeEach(async () => {
				await wrapper.setData({
					adminConfigStatus: false,
					state: { user_name: 'test', token: '123' },
				})
			})
			it('should set proper props to the oauth connect component', () => {
				expect(wrapper.find(oAuthButtonSelector).exists()).toBeTruthy()
				expect(wrapper.find(oAuthButtonSelector).props()).toMatchObject({
					requestUrl: 'https://nextcloud.com/',
					adminConfigStatus: false,
				})
			})
		})
	})
})
