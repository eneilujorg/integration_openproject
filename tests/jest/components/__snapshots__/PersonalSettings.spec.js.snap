// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PersonalSettings.vue oAuth user settings admin and user should be able to enable and disable the setting simultaneously 1`] = `
<div class="openproject-prefs--form">
  <div class="form-check-input"><input id="openproject-prefs--link" type="checkbox" class="checkbox" value="true"> <label for="openproject-prefs--link">
      Enable navigation link
    </label>
    <p class="user-setting-description">Displays a link to your OpenProject instance in the Nextcloud header.</p>
  </div>
  <div class="form-check-input"><input id="openproject-prefs--u-search" type="checkbox" class="checkbox" value="true"> <label for="openproject-prefs--u-search">
      Enable unified search for tickets
    </label>
    <p class="user-setting-description">Allows you to search OpenProject work packages via the universal search bar in Nextcloud.</p>
    <p class="openproject-prefs--hint"><span aria-hidden="true" role="img" class="material-design-icon information-variant-icon"><svg fill="currentColor" width="24" height="24" viewBox="0 0 24 24" class="material-design-icon__svg"><path d="M13.5,4A1.5,1.5 0 0,0 12,5.5A1.5,1.5 0 0,0 13.5,7A1.5,1.5 0 0,0 15,5.5A1.5,1.5 0 0,0 13.5,4M13.14,8.77C11.95,8.87 8.7,11.46 8.7,11.46C8.5,11.61 8.56,11.6 8.72,11.88C8.88,12.15 8.86,12.17 9.05,12.04C9.25,11.91 9.58,11.7 10.13,11.36C12.25,10 10.47,13.14 9.56,18.43C9.2,21.05 11.56,19.7 12.17,19.3C12.77,18.91 14.38,17.8 14.54,17.69C14.76,17.54 14.6,17.42 14.43,17.17C14.31,17 14.19,17.12 14.19,17.12C13.54,17.55 12.35,18.45 12.19,17.88C12,17.31 13.22,13.4 13.89,10.71C14,10.07 14.3,8.67 13.14,8.77Z"><!----></path></svg></span>
      Warning, everything you type in the search bar will be sent to your OpenProject instance.
    </p>
  </div>
</div>
`;

exports[`PersonalSettings.vue oAuth user settings admin and user should be able to enable and disable the setting simultaneously 2`] = `
<div class="openproject-prefs--form">
  <div class="form-check-input"><input id="openproject-prefs--link" type="checkbox" class="checkbox" value="false"> <label for="openproject-prefs--link">
      Enable navigation link
    </label>
    <p class="user-setting-description">Displays a link to your OpenProject instance in the Nextcloud header.</p>
  </div>
  <div class="form-check-input"><input id="openproject-prefs--u-search" type="checkbox" class="checkbox" value="true"> <label for="openproject-prefs--u-search">
      Enable unified search for tickets
    </label>
    <p class="user-setting-description">Allows you to search OpenProject work packages via the universal search bar in Nextcloud.</p>
    <p class="openproject-prefs--hint"><span aria-hidden="true" role="img" class="material-design-icon information-variant-icon"><svg fill="currentColor" width="24" height="24" viewBox="0 0 24 24" class="material-design-icon__svg"><path d="M13.5,4A1.5,1.5 0 0,0 12,5.5A1.5,1.5 0 0,0 13.5,7A1.5,1.5 0 0,0 15,5.5A1.5,1.5 0 0,0 13.5,4M13.14,8.77C11.95,8.87 8.7,11.46 8.7,11.46C8.5,11.61 8.56,11.6 8.72,11.88C8.88,12.15 8.86,12.17 9.05,12.04C9.25,11.91 9.58,11.7 10.13,11.36C12.25,10 10.47,13.14 9.56,18.43C9.2,21.05 11.56,19.7 12.17,19.3C12.77,18.91 14.38,17.8 14.54,17.69C14.76,17.54 14.6,17.42 14.43,17.17C14.31,17 14.19,17.12 14.19,17.12C13.54,17.55 12.35,18.45 12.19,17.88C12,17.31 13.22,13.4 13.89,10.71C14,10.07 14.3,8.67 13.14,8.77Z"><!----></path></svg></span>
      Warning, everything you type in the search bar will be sent to your OpenProject instance.
    </p>
  </div>
</div>
`;

exports[`PersonalSettings.vue oAuth user settings should be disabled if the admin has not enabled the settings 1`] = `
<div class="openproject-prefs--form">
  <checkbox-stub inputid="openproject-prefs--link" label="Enable navigation link"></checkbox-stub>
  <checkbox-stub inputid="openproject-prefs--u-search" label="Enable unified search for tickets"></checkbox-stub>
</div>
`;

exports[`PersonalSettings.vue oAuth user settings should be enabled if the admin has enabled the settings 1`] = `
<div class="openproject-prefs--form">
  <checkbox-stub value="true" inputid="openproject-prefs--link" label="Enable navigation link"></checkbox-stub>
  <checkbox-stub value="true" inputid="openproject-prefs--u-search" label="Enable unified search for tickets"></checkbox-stub>
</div>
`;
